File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-2.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        

        maxA = 0
        stack = []
        for i,h in enumerate(heights):
            newIdx = i
            while stack and stack[-1][1] > h:
                newIdx, hght = stack.pop()
                maxA = max(maxA, hght * (i-newIdx))
            stack.append((newIdx,h))

        for i,h in stack:
            maxA = max(maxA,  h * (len(heights)-i))

        return maxA

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-2.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> window(256);
        int res = 0;
        int N = s.size();

        int L = 0, R = 0;
        while (R < N) {
            while (window[s[R]]) {
                window[s[L]]--;
                L++;
            }
            window[s[R]]++;
            res = max(res,R-L+1);
            R++;
        }
        return res;
    }
};
